hasher-matcher-actioner/hmalib/common/writebacker_models.py

```
@@ -0,0 +1,240 @@
+        writebacks that can be taken as a mapping from writback type to writebacker. The
+        type should be same as the type of is passed in the WritebackLabel passed to the writebacker
+        """
+        raise NotImplementedError
+
+    @classmethod
+    @lru_cache(maxsize=None)
+    def get_writebacker_for_source(cls, source: str) -> t.Optional["Writebacker"]:
+        if cls.__name__ != "Writebacker":
+            raise ValueError(
+                "get_writebacker_for_source can only be called from the Writebacker class directly. eg Writebacker().get_writebacker_for_source"
+            )
+
+        print("subclasses", cls.__subclasses__())
+
+        sources_to_writebacker_cls = {
+            writebacker_cls().source: writebacker_cls
+            for writebacker_cls in cls.__subclasses__()
+        }
+
+        if source not in sources_to_writebacker_cls.keys():
+            return None
+        return sources_to_writebacker_cls[source]()
+
+    def writeback_is_enabled(self) -> bool:
+        """
+        Users can switch on/off writebacks either globally or for individual sources
+        """
+        raise NotImplementedError
+
+    @property
+    def writeback_type(self) -> WritebackLabel:
+        """
+        The writeback label for when this action should be performed (eg SawThisTooWritebackLabel())
+        """
+        raise NotImplementedError
+
+    def _writeback_impl(self, writeback_message: WritebackMessage) -> str:
+        raise NotImplementedError
+
+    def perform_writeback(self, writeback_message: WritebackMessage) -> str:
+        writeback_options = self.writeback_options()
+        writeback_to_perform = writeback_message.writeback_label.value
+        if writeback_to_perform not in writeback_options:
+            return (
+                "Could not find writebacker for source "
+                + self.source
+                + " that can perform writeback "
+                + writeback_to_perform
+            )
```

Will this return raise an error or log something somewhere? I think if this kind of situation exists we'd want to surface an error in the logs, right?

---

Adding more logging through out this file

---

But as an error or just info?